src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerConfigOptions.cs

```diff
@@ -23,18 +22,27 @@ public abstract class AnalyzerConfigOptions
         public abstract bool TryGetValue(string key, out string value);
     }
 
-    internal sealed class CompilerAnalyzerConfigOptions : AnalyzerConfigOptions
+    public abstract class DictionaryBackedAnalyzerConfigOptions : AnalyzerConfigOptions
     {
-        public static CompilerAnalyzerConfigOptions Empty { get; } = new CompilerAnalyzerConfigOptions(EmptyDictionary);
-
         private readonly ImmutableDictionary<string, string> _backing;
 
-        public CompilerAnalyzerConfigOptions(ImmutableDictionary<string, string> properties)
+        public DictionaryBackedAnalyzerConfigOptions(ImmutableDictionary<string, string> properties)
         {
             _backing = properties;
         }
 
         public override bool TryGetValue(string key, out string value) => _backing.TryGetValue(key, out value);
 
+        public IReadOnlyDictionary<string, string> AsReadOnlyDictionary() => _backing;
```

Comment:
@agocke Didn't we want to avoid exposing this as a IReadOnlyDictionary for the most part? Because now this exposes enumeration which is Complicated™?

Reply:
It also seems that if we _did_ want to encode an API for this, it should be an interface, so other implementations could expose this without forcing their inheritence hierarchy.

Reply:
Getting NamingStyleOptions expects the ability to enumerate all the keys defined in the .editorconfig. My dictionary wrapper for AnalyzerConfigOptions was insufficient since I was unable to provide that capability.

Reply:
@agocke @sharwell Do either of you remember discussing how to account for NamingStyleOptions when designing the .editorconfig support?

Reply:
After talking to @agocke, an enumerator was not offered because the expectation that the IDE would implement the AnalyzerConfigOptions in a way that merged the config options with the Workspace OptionSet. At the time, the OptionSet was not part of the Solution snapshot so creating a reliable enumeration wasn't possible. Now that OptionSet is tied to a snapshot we should be able to enumerate all the options without fear of things changing underneath us.